### PR TITLE
Cross-Lingual objectives (XLM) and multilingual (many-many) support for Megatron-NMT

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds the following

1. XLM + UL2-based datasets for BERT and SeqSeq
2. Multilingual support for MegatronNMT (x -> En, En -> x and x <-> y).

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by aklife97** [comment]
adding comment to remove staleness, publishing review soon.

**Review by aklife97 — APPROVED** [review]
LGTM, great PR!!

File: examples/nlp/machine_translation/conf/aayn_base_megatron.yaml [hunk 1]
```diff
@@ -93,35 +93,29 @@ model:
...
+    sampling_ratios:
+      x-masking: 0.17 # Extreme span masking task selection probability (either large spans or large masking prob)
+      r-masking: 0.17 # T5-style random span masking task selection probability
+      s-masking: 0.16 # Prefix-LM task selection probability
+      nmt: 0.5 # NMT selection probability
     micro_batch_size: ${model.micro_batch_size}
     global_batch_size: ${model.global_batch_size}
     # config for preprocessing training data and creating a tarred datset automatically
-    tar_file_prefix: parallel # prefix for tar file names
-    tar_files: null # if data has already been preprocessed (rest of config ignored)
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/machine_translation/conf/aayn_base_megatron.yaml]: Hey! I removed tarred dataset support in an earlier PR - I had just pulled that branch into this branch. PRs here https://github.com/NVIDIA/NeMo/pull/5142 & https://github.com/NVIDIA/NeMo/pull/5065. The old tarred datasets weren't compatible with gradacc and the apex fwd/bwd functions and I had to really hack around them earlier.

**Review by aklife97 — DISMISSED** [review]
LGTM! Just some questions:

File: nemo/collections/nlp/data/language_modeling/megatron/xlm_dataset.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+        favor_long_ngrams: bool = False,
+        permutation: bool = False,
+        geometric_dist: bool = True,
+        tokenizer_type: str = "wordpiece",
+    ):
+        # Determine which task to perform - NMT/T5/UL2 based on sampling ratios.
+        task = np_rng.choice(list(sampling_ratios.keys()), p=list(sampling_ratios.values()))
+
+        # Potentially swap src, tgt with a 50% chance to avoid learning associations based on position in the sequence.
+        swap_src_tgt = np_rng.randint(0, 2)
```
Comment (aklife97) [inline nemo/collections/nlp/data/language_modeling/megatron/xlm_dataset.py:352]: (reference)

**Review by aklife97 — APPROVED** [review]

**Pull request merged**